src/publishProject/projectUtils.ts

```diff
@@ -143,6 +163,63 @@ export function validateSqlServerPortNumber(port: number): boolean {
     return Number.isInteger(port) && port >= 1 && port <= constants.MAX_PORT_NUMBER;
 }
 
+/**
+ * Retrieves and filters SQL Server container tags based on project target version.
+ * Used by Publish Project dialog to provide granular tag selection.
+ *
+ * This follows the same filtering logic as ADS (Azure Data Studio):
+ * - If target version is known: filter to show versions >= target version year
+ * - If target version is unknown: fallback to max available version year (like ADS)
+ *
+ * @param targetVersion - The SQL Server version (e.g., "160" for SQL Server 2022)
+ * @returns Sorted array of tags filtered by version from deployment UI versions
+ */
+export async function getSqlServerContainerTagsForTargetVersion(
+    targetVersion?: string,
+): Promise<FormItemOptions[]> {
+    try {
+        // Get the deployment UI versions first
+        const deploymentVersions = await getSqlServerContainerVersions();
+        if (!deploymentVersions || deploymentVersions.length === 0) {
+            return [];
+        }
+
+        // Extract years from all available versions to find the maximum
+        const availableYears = deploymentVersions
```

Comment:
How about building a Map<int, string> where the key is the year and the value is the full tag?  I think that would clean this up.